### PR TITLE
fabric: fix vendored lib "cryptography" linking to getentropy()

### PR DESCRIPTION
**Pull request by ploxiln**
paramiko 2.x uses the "cryptography" module, which can have issues linking to _getentropy from libSystem.B.dylib on OS X 10.11

fix for #8044

-----

- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by ilovezfs** [comment]
would prefer we actually fix this and report it.

**Comment by ploxiln** [comment]
I'm open to suggestions. A tricky part is that I'm running OS X 10.11.6 and it works for me. But the bottle is broken. Do you know what toolchain the CI build hosts are running? Mine for reference:

```
$ clang --version
Apple LLVM version 8.0.0 (clang-800.0.42.1)
Target: x86_64-apple-darwin15.6.0
```

**Comment by ilovezfs** [comment]
what's your `brew config` output?

**Comment by ploxiln** [comment]
```
$ brew config
HOMEBREW_VERSION: 1.1.5-20-g0027ded5aa
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: 0027ded5aa6195dcc153ecf0d3d3b05e99f339af
Last commit: 24 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 73f58daede6ceb8a3a74c1eafd4ed133b3d9a867
Core tap last commit: 70 minutes ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_REPOSITORY: /usr/local/Homebrew
HOMEBREW_CELLAR: /usr/local/Cellar
HOMEBREW_BOTTLE_DOMAIN: https://homebrew.bintray.com
CPU: quad-core 64-bit haswell
Homebrew Ruby: 2.0.0-p648
Clang: 8.0 build 800
Git: 2.11.0 => /usr/local/bin/git
Perl: /usr/bin/perl
Python: /usr/bin/python
Ruby: /Users/pierce/.rbenv/shims/ruby => /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby
Java: 1.8.0_71
macOS: 10.11.6-x86_64
Xcode: N/A
CLT: 8.2.0.0.1.1480973914
X11: 2.7.9 => /opt/X11
```

**Comment by ilovezfs** [comment]
```
diff --git a/Formula/fabric.rb b/Formula/fabric.rb
index bd52a1d..31541ce 100644
--- a/Formula/fabric.rb
+++ b/Formula/fabric.rb
@@ -69,6 +69,11 @@ class Fabric < Formula
     ENV.prepend_create_path "PYTHONPATH", libexec/"vendor/lib/python2.7/site-packages"
     resources.each do |r|
       r.stage do
+        if r.name == "cryptography" && MacOS.version < :sierra && DevelopmentTools.clang_build_version >= 800
+          inreplace "src/_cffi_src/openssl/src/osrandom_engine.h",
+            "#elif defined(BSD) && defined(SYS_getentropy)",
+            "#elif defined(BSD) && defined(undefined_gibberish)"
+        end
         system "python", *Language::Python.setup_install_args(libexec/"vendor")
       end
     end
```
fixes it.

**Comment by ploxiln** [comment]
Great idea. I pushed up a slight variant of that ... because I was suspicious of the clang_build_version check. I thought the latest version did *not* have the problem, only *previous* versions should be patched? Anyway, no need to check dev tools version really, can just make damn sure `getentropy()` is not used on `MacOS.version < :sierra`

**Comment by ploxiln** [comment]
I did not experience this failure myself when building, so the only way for me to really check this fix worked (didn't have e.g. a typo in the replace) would be to download/install the bottle that's generated from this PR ...

**Comment by ilovezfs** [comment]
You can check the binary with `nm` for the naughty symbol.

**Comment by ploxiln** [comment]
I can check the binary in the bottle built by jenkins with nm, but the one built on my system has never had the symbol, and will never have it. That's why it works on my system. Even though I'm running MacOS 10.11 and have clang build 800 installed. My `/usr/include/sys/syscall.h` does not have `SYS_getentropy`.

```
$ grep gete /usr/include/sys/syscall.h 
#define	SYS_geteuid        25
#define	SYS_getegid        43
$ nm /usr/local/Cellar/fabric/1.13.1/libexec/vendor/lib/python2.7/site-packages/cryptography/hazmat/bindings/_openssl.so | grep getentropy
$ brew fetch --force-bottle fabric
==> Downloading https://homebrew.bintray.com/bottles/fabric-1.13.1.el_capitan.bottle.tar.gz
######################################################################## 100.0%
Downloaded to: /Users/pierce/Library/Caches/Homebrew/fabric-1.13.1.el_capitan.bottle.tar.gz
SHA256: 858ac38f6bfb7badd0321718e0c100bb1666358c314d955ded8f18ac2cc09c25
$ tar -xf /Users/pierce/Library/Caches/Homebrew/fabric-1.13.1.el_capitan.bottle.tar.gz
$ nm fabric/1.13.1/libexec/vendor/lib/python2.7/site-packages/cryptography/hazmat/bindings/_openssl.so | grep getentropy
                 U _getentropy
```

My latest guess is that the problem is XCode toolchain vs "Command Line Tools for 10.11", I have the latter, while people who have the former (incl. the build bots) get something that's common/shared between 10.11 and 10.12.

Anyway, I'll have to wait for the jenkins jobs to build the bottle and then figure out how to download it so the fix can be verified before merge.

**Comment by ilovezfs** [comment]
Yes, you need Xcode to test it.

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by ploxiln** [comment]
in addition to @stxml testing this branch and confirming the fix in #8044, I also downloaded the bottle just created by the build bot and verified

```
[pierce@plo-pro tmp]$ wget -q https://bot.brew.sh/job/Homebrew%20Core/13464/version=el_capitan/artifact/fabric-1.13.1_1.el_capitan.bottle.tar.gz
[pierce@plo-pro tmp]$ tar -xf fabric-1.13.1_1.el_capitan.bottle.tar.gz
[pierce@plo-pro tmp]$ nm fabric/1.13.1_1/libexec/vendor/lib/python2.7/site-packages/cryptography/hazmat/bindings/_openssl.so | grep getentropy
[pierce@plo-pro tmp]$ 
```

... full Xcode install successfully put off yet again ;)

**Comment by ilovezfs** [comment]
>... full Xcode install successfully put off yet again ;)

Resistance is futile.

**Comment by ploxiln** [comment]
this probably no longer needs "do not merge" label ...

**Comment by ilovezfs** [comment]
Agreed. However, the bug needs to be reported upstream with the report link, date, and description before it :ship:

**Comment by ilovezfs** [comment]
@ploxiln here are examples of what I mean: https://gist.github.com/ilovezfs/62a9eb715dc044f7202cbf86caaf713b

**Comment by ploxiln** [comment]
https://github.com/pyca/cryptography/issues/3332

**Comment by ploxiln** [comment]
ok, I see what you mean, updated comment in formula

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
@ploxiln great work! We can now :shipit: when it's :green_apple:

**Comment by ploxiln** [comment]
I appreciate all the help, thanks @ilovezfs

**Comment by ilovezfs** [comment]
@ploxiln shipped! Thanks for your contribution to Homebrew. :balloon:

**Pull request closed without merging**
